### ISSUE TITLE
Spaces between circles in GDPaint

### ISSUE THREAD

**Issue by michaldev** [godotengine/godot-demo-projects#846]
Hi. Why in your demo I have spaces between circles? How can i fix it?

https://user-images.githubusercontent.com/5132385/219780927-0dbb08e1-91b8-44ab-b30e-0ef6fa254265.mov

**Comment by Calinou** [comment]
This is because the Bresenham line drawing algorithm needs to be implemented in the demo, placing a circle every pixel between the last two reported input events.

**Comment by michaldev** [comment]
Something like Line2D (points between mouse movement)?

**Comment by Calinou** [comment]
> Something like Line2D (points between mouse movement)?

No, the Bresenham algorithm is something you can implement easily in GDScript.

**Comment by michaldev** [comment]
Why something like this is bad method?
```
func _draw():
	var index = 0
	for point in _points:
		if (len(_points) != index+1):
			var brush_size = 10
			draw_circle(point, brush_size, Color.NAVY_BLUE)
			draw_line(point, _points[index+1], Color.NAVY_BLUE, brush_size*2)
		index += 1
```

**Comment by Calinou** [comment]
> Why something like this is bad method (it works)?

Can you show a video/GIF of drawing lines with fast mouse motion with this method? It might be good enough.

**Comment by michaldev** [comment]
https://user-images.githubusercontent.com/5132385/219813333-5d739000-0bdd-4a25-a786-dee30092ceda.mov

**Comment by michaldev** [comment]
Uploading Nagranie z ekranu 2023-02-18 o 00.02.51.mov…

**Comment by Calinou** [comment]
@michaldev Looks good! Please open a pull request :slightly_smiling_face:

**Comment by michaldev** [comment]
I thought about it and I think it's a bad method. You will not be able to create an eraser tool.

**Comment by michaldev** [comment]
I tested this solution for a longer time and I think it's not very good - when there are more drawn lines, significant delays appear between the mouse cursor and the currently drawn line.

**Comment by Calinou** [comment]
> I tested this solution for a longer time and I think it's not very good - when there are more drawn lines, significant delays appear between the mouse cursor and the currently drawn line.

The way the demo is designed allows for more flexibility, but it was never intended to be as fast as possible.

**Comment by dalexeev** [comment]
> This is because [Bresenham's line algorithm](https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm) needs to be implemented in the demo

* See godotengine/godot#74714.

**Comment by Zylann** [comment]
I wonder why bresenham's line algorithm would be so useful here? It's a circular brush. Bresenham only gives you a series of positions for a single-pixel line, allocating a long array as well. Using it many times to fill a thicker line (especially as shown earlier) doesn't seem great when the GPU can draw with polygons. Using it to draw many circles wouldn't be good either. Assuming this is painting on a Viewport, https://github.com/godotengine/godot-demo-projects/issues/846#issuecomment-1435347465 would work better, although for opaque colors only. It won't do well with transparent. The algorithm Line2D uses would work best out of the box, but since it wasn't used in `draw_line`, to be used here it requires code changes compared to a function call (need to instantiate a node). But it depends what the demo wants to show of course.
